// duice/dist/ArrayElement.js
import { Element } from "./Element";
import { Observable } from "./Observable";
import { Event } from "./event/Event";
/**
 * Array Element
 */
export declare class ArrayElement<T extends HTMLElement> extends Element<T, object[]> {
    loop: string;
    hierarchy: string;
    editable: boolean;
    selectedItemClass: string;
    slot: HTMLSlotElement;
    itemHtmlElements: HTMLElement[];
    /**
     * Constructor
     * @param htmlElement html element
     * @param bindData bind data
     * @param context context
     */
    constructor(htmlElement: T, bindData: object[], context: object);
    /**
     * Renders
     */
    render(): void;
    /**
     * Creates item html element
     * @param index index
     * @param object object
     * @param context context
     */
    createItemHtmlElement(index: number, object: object, context: object): void;
    /**
     * Updates
     * @param observable observable
     * @param event event
     */
    update(observable: Observable, event: Event): void;
}
